kong/clustering/config_helper.lua

```diff
@@ -1,132 +1,147 @@
 local constants = require("kong.constants")
 local declarative = require("kong.db.declarative")
+local tablepool = require("tablepool")
+local isempty = require("table.isempty")
+local isarray = require("table.isarray")
+local nkeys = require("table.nkeys")
+local buffer = require("string.buffer")
+
 
 local tostring = tostring
 local assert = assert
 local type = type
 local error = error
 local pairs = pairs
 local ipairs = ipairs
-local concat = table.concat
 local sort = table.sort
-
-local isempty = require("table.isempty")
-local isarray = require("table.isarray")
-local nkeys = require("table.nkeys")
-local new_tab = require("table.new")
-
 local yield = require("kong.tools.utils").yield
+local fetch_table = tablepool.fetch
+local release_table = tablepool.release
+
 
 local ngx_log = ngx.log
 local ngx_null = ngx.null
 local ngx_md5 = ngx.md5
 local ngx_md5_bin = ngx.md5_bin
 
+
 local ngx_DEBUG = ngx.DEBUG
 
+
 local DECLARATIVE_EMPTY_CONFIG_HASH = constants.DECLARATIVE_EMPTY_CONFIG_HASH
 local _log_prefix = "[clustering] "
 
 
 local _M = {}
 
 
-local function to_sorted_string(value)
+local function to_sorted_string(value, o)
   yield(true)
 
-  if value == ngx_null then
-    return "/null/"
-  end
-
-  local t = type(value)
-  if t == "string" or t == "number" then
-    return value
-  end
-
-  if t == "boolean" then
-    return tostring(value)
+  if #o > 512 then
+    o:set(ngx_md5_bin(o:tostring()))
   end
 
-  if t == "table" then
-    if isempty(value) then
-      return "{}"
-    end
-
-    if isarray(value) then
-      local count = #value
-      if count == 1 then
-        return to_sorted_string(value[1])
-      end
-
-      if count == 2 then
-        return to_sorted_string(value[1]) .. ";" ..
-               to_sorted_string(value[2])
-      end
-
-      if count == 3 then
-        return to_sorted_string(value[1]) .. ";" ..
-               to_sorted_string(value[2]) .. ";" ..
-               to_sorted_string(value[3])
-      end
-
-      if count == 4 then
-        return to_sorted_string(value[1]) .. ";" ..
-               to_sorted_string(value[2]) .. ";" ..
-               to_sorted_string(value[3]) .. ";" ..
-               to_sorted_string(value[4])
-      end
-
-      if count == 5 then
-        return to_sorted_string(value[1]) .. ";" ..
-               to_sorted_string(value[2]) .. ";" ..
-               to_sorted_string(value[3]) .. ";" ..
-               to_sorted_string(value[4]) .. ";" ..
-               to_sorted_string(value[5])
-      end
-
-      local i = 0
-      local o = new_tab(count < 100 and count or 100, 0)
-      for j = 1, count do
-        i = i + 1
-        o[i] = to_sorted_string(value[j])
+  if value == ngx_null then
+    o:put("/null/")
+
+  else
+    local t = type(value)
+    if t == "string" or t == "number" then
```

Comment:
since we are using string buffer, i think we can also just skip type assertions and let it do the job. previously we do this because {} could be from a table or from a string "{}". but they are already encoded differently in string.buffer.

Reply:
@fffonion only `string`, `number` and `object` with `__tostring` are allowed with `:put`